src/PrimaryGeneratorAction.cc

```diff
@@ -39,132 +41,134 @@ PrimaryGeneratorAction::~PrimaryGeneratorAction() { delete fParticleGun; }
-        if (generator_type ==
-            g4_metadata_parameters::generator_types::CUSTOM) {  // Generator type "file": no randomisation
-            static int nEvts = 0;
-            rndCollection = nEvts++;
-        } else
-            rndCollection = (Int_t)(G4UniformRand() * nCollections);
-
-        restG4Metadata->ReadParticleCollection(rndCollection);
+void PrimaryGeneratorAction::GeneratePrimaries(G4Event* geant4_event) {
+    if (restG4Metadata->GetVerboseLevel() >= REST_Debug) {
+        cout << "DEBUG: Primary generation" << endl;
     }
+    // We have to initialize here and not in start of the event because
+    // GeneratePrimaries is called first, and we want to store event origin and
+    // position inside
+    // we should have already written the information from previous event to disk
+    // (in endOfEventAction)
+    restG4Event->Initialize();
 
-    Int_t nParticles = restG4Metadata->GetNumberOfPrimaries();
+    for (int i = 0; i < restG4Metadata->GetNumberOfSources(); i++) {
+        restG4Metadata->GetParticleSource(i)->Update();
+    }
 
-    if (generator_type !=
-        g4_metadata_parameters::generator_types::CUSTOM)  // Except for generator "custom"...
-        // ...Position is common for all particles
-        SetParticlePosition();
+    Int_t nParticles = restG4Metadata->GetNumberOfSources();
 
-    for (int j = 0; j < nParticles; j++) {
-        if (generator_type == g4_metadata_parameters::generator_types::CUSTOM)  // Generator type "custom"...
-            // ...Get position from particle collection
-            SetParticlePosition(j);
+    // Set the particle(s)' position, multiple particles generated from multiple
+    // sources shall always have a same origin
+    SetParticlePosition();
 
-        // ParticleDefinition should be always declared first (after position).
-        SetParticleDefinition(j);
+    for (int i = 0; i < restG4Metadata->GetNumberOfSources(); i++) {
+        vector<TRestGeant4Particle> particles = restG4Metadata->GetParticleSource(i)->GetParticles();
+        for (auto p : particles) {
+            // ParticleDefinition should be always declared first (after position).
+            SetParticleDefinition(i, p);
 
-        // Particle Direction must be always set before energy
-        SetParticleEnergy(j);
+            // Particle Direction must be always set before energy
+            SetParticleEnergy(i, p);
 
-        SetParticleDirection(j);
+            SetParticleDirection(i, p);
 
-        fParticleGun->GeneratePrimaryVertex(geant4_event);
+            fParticleGun->GeneratePrimaryVertex(geant4_event);
+        }
     }
 }
 
 //_____________________________________________________________________________
-G4ParticleDefinition* PrimaryGeneratorAction::SetParticleDefinition(int n) {
-    string particle_name = (string)restG4Metadata->GetParticleSource(n).GetParticleName();
+G4ParticleDefinition* PrimaryGeneratorAction::SetParticleDefinition(Int_t n, TRestGeant4Particle p) {
```

Comment:
Would be better to pass TRestGeant4Particle as reference

`PrimaryGeneratorAction::SetParticleDefinition(Int_t n, const TRestGeant4Particle &p)`